### DISCUSSION TITLE
在Umi4中使用qiankun插件，子应用如何向父应用传递数据

### DISCUSSION THREAD

**Discussion by GuanJdoJ** [umijs/umi#11846] [Q&A]
有这样一个场景：两应用独立时都使用自定义菜单组件MyMenu，当做微应用时，希望子应用自身不展示菜单，能把自己的菜单数据传给父应用，父应用合并处理菜单数据后放到自身的MyMenu中，所以需要能够能让子向父传递数据。

尝试过在子应用中用@@qiankunStateFromMaster中的setGlobalState回调传值，没有生效。

请问有什么好的方案吗？

**Comment by fz6m** [comment]
我试了下，子应用使用父应用传过来的 `setState` 方法是可以的，父应用 `useEffect` 是可以监听到 state 的变化的。

如有问题，需要给一个不生效的最小复现。

**Reply by GuanJdoJ** [reply]
父应用
```tsx
//app.tsx
export function useQiankunStateForSlave() {
  const [globalState, setGlobalState] = useState<any>({
    slogan: 'Hello MicroFrontend',
  });

  useEffect(() => {
    console.log('父-globalState', globalState);
  }, [globalState]);

  return {
    globalState,
    setGlobalState,
  };
}
```

子应用
```tsx
//src/components/MyLayout.tsx
import { useModel } from 'umi';

export default function Layout() {
  const masterProps = useModel('@@qiankunStateFromMaster');

  if(masterProps && masterProps.setGlobalState) {
    masterProps.setGlobalState({
      slogan: '你好'
    })
  }
  
//...
}
```

我是这样写的不生效，而且控制台还会有警告
<img width="554" alt="image" src="https://github.com/umijs/umi/assets/30386473/afc601e0-af39-4f2d-832b-552dbff96abb">


然后我还有一个问题，如何在父应用组件中获取globalState数据呢，我这样也没成功
```tsx
//父应用 /src/components/SomeComp.tsx
import { useQiankunStateForSlave } from '@/app';

export default function SomeComp() {
  const { globalState } = useQiankunStateForSlave();
  console.log('父-Layout-globalState', globalState);
}
```

**Reply by fz6m** [reply]
不可以在组件中直接调用 `masterProps.setGlobalState` ，此时有可能还没接收到，或者导致无限 set state 组件无限刷新，需要放到实际的逻辑触发时机或 `useEffect` 中调用。

如果要把 `export const useQiankunStateForSlave` 中的 `globalState` 分享到全局，可以使用 react 全局数据的库来创建数据，比如 [`jotai`](https://github.com/pmndrs/jotai) 或 [valtio](https://umijs.org/docs/max/valtio) 。

**Reply by GuanJdoJ** [reply]
感谢您的回答，我调整了下代码如下：
- 父应用
```tsx
//app.tsx
export function useQiankunStateForSlave() {
  const [globalState, setGlobalState] = useState<any>({
    slogan: 'Hello MicroFrontend',
  });

  const [msg, setMsg] = useState('');
  const callback = (data: { name: string; msg: string }) => {
    console.log('父-data', data);
    setMsg(data.msg);
  };

  useEffect(() => {
    console.log('父-globalState', globalState);
  }, [globalState]);

  return {
    globalState,
    setGlobalState,
    callback,
    msg,
  };
}
```
```tsx
//父应用 /src/components/SomeComp.tsx
import { useModel } from 'umi';

export default function SomeComp() {
  const { msg, globalState } = useModel('@@qiankunStateForSlave');
  console.log('父-Layout-globalState', msg, globalState);
}
```

- 子应用
```tsx
//src/components/MyLayout.tsx
import { useModel } from 'umi';
import { Button } from 'antd';

export default function Layout() {
  const masterProps = useModel('@@qiankunStateFromMaster');

  const handlePostMsgToMaster = () => {
    const { setGlobalState, callback } = masterProps || {};
    if (callback) {
      callback({
        name: 'test',
        msg: 'test',
      });
    }
    if (setGlobalState) {
      setGlobalState({
        slogan: '你好',
      });
    }
  };

  return (
    <div>
      <Button onClick={handlePostMsgToMaster}>向父应用发消息</Button>
    </div>
  )
}
```

1.我调整了子应用中`setGlobalState`的调用时机，仍然无法通过`setGlobalState`来改变数据，如下图所示，slogan没变且提示warnning，但是通过传callback函数可以实现，如果我此时`setGlobalState`的用法还是不对麻烦您给我个demo；（父子应用`umi@4.0.86` `@umijs/plugins@4.0.86`）
2.我尝试在父应用中使用`useModel('@@qiankunStateForSlave')`来获取数据，这样做是能成功的，如下图`父-data`打印结果显示接受到了回调数据，这是正确可靠的用法吗？
<img width="560" alt="image" src="https://github.com/umijs/umi/assets/30386473/1db7f3c2-debe-4560-8979-678e21b7a549">


我还想问一个问题：
3.在父应用没有访问子应用路由时，或者子应用没有加载时，父应用有办法获取子应用的数据吗（回到我最上面提问的场景，父应用需要在没访问子应用时就获取到子应用的菜单数据，这样才能完整显示所有菜单）？

**Reply by fz6m** [reply]
1. 我用最新版本的 umi 是没有你说的第 1 个问题的，父应用传递下去的 setState 可以正常调用，父应用的 `useEffect` 可以监听到。如果你的某个应用是 umi 而不是 max ，要把 `qiankun` 和 `model` 插件都启用，如还有问题，需要给一个最小复现仓库。

2. 不要使用 `@@qiankunStateForSlave` 这是你自己声明的 `useQiankunStateForSlave` ，不需要在父应用自己调用。

3. 如果希望更早的获取子应用信息，建议做一个配置中心或者全局配置统一动态下发，可以把菜单配置统一到某个接口或者配置中心去，因为子应用的加载还是有些延后的，另外可以考虑用 [`<MicroApp />`](https://umijs.org/docs/max/micro-frontend#microapp--%E7%BB%84%E4%BB%B6%E5%BC%95%E5%85%A5%E5%AD%90%E5%BA%94%E7%94%A8) 来手动渲染对应的子应用，这样可以想办法更早提前启动子应用。
